tests/unit/test_client.py

```diff
@@ -254,225 +259,360 @@ def long_call(request, uri, headers):
+        self.token_server = token_server
+        self.tokens = tokens
+        self.sample_post_response_data = sample_post_response_data
 
-@mock.patch("requests.Session.post")
-def test_oauth2_authentication_missing_headers(http_session_post):
-    http_session = requests.Session()
-    oauth_test = OAuthTestReq("post", OAUTH_SERVER_URL_NO_HEADER)
-    http_session_post.return_value = oauth_test.get_statement_post_response()
-    oauth = _OAuth2TokenBearer(http_session, oauth2_test_url_handler)
+    def __call__(self, request, uri, response_headers):
+        authorization = request.headers.get("Authorization")
+        if authorization and authorization.replace("Bearer ", "") in self.tokens:
+            return [200, response_headers, json.dumps(self.sample_post_response_data)]
+        return [401, {'Www-Authenticate': f'Bearer x_redirect_server="{self.redirect_server}", '
+                                          f'x_token_server="{self.token_server}"',
+                      'Basic realm': '"Trino"'}, ""]
 
-    with pytest.raises(trino.exceptions.TrinoAuthError) as exp:
-        statement_req = oauth_test.get_statement_request()
-        oauth(statement_req)
 
-    assert str(exp.value) == "Error: header WWW-Authenticate not available in the response."
+class GetTokenCallback:
+    def __init__(self, token_server, token, attempts=1):
+        self.token_server = token_server
+        self.token = token
+        self.attempts = attempts
 
+    def __call__(self, request, uri, response_headers):
+        self.attempts -= 1
+        if self.attempts < 0:
+            return [404, response_headers, "{}"]
+        if self.attempts == 0:
+            return [200, response_headers, f'{{"token": "{self.token}"}}']
+        return [200, response_headers, f'{{"nextUri": "{self.token_server}"}}']
 
-@mock.patch("requests.Session.get")
-@mock.patch("requests.Session.post")
-def test_oauth2_authentication_fail_token_server(http_session_post, http_session_get):
-    http_session = requests.Session()
-    oauth_test = OAuthTestReq("post", OAUTH_SERVER_URL_FAIL_SERVER)
-    http_session_post.return_value = oauth_test.get_statement_post_response()
-    oauth = _OAuth2TokenBearer(http_session, oauth2_test_url_handler)
-    http_session_get.side_effect = oauth_test.get_token_url_response()
 
-    statement_req = oauth_test.get_statement_request()
-    oauth(statement_req)
+@pytest.mark.parametrize("attempts", [1, 3, 5])
+@httprettified
+def test_oauth2_authentication_flow(attempts, sample_post_response_data):
+    token = str(uuid.uuid4())
+    challenge_id = str(uuid.uuid4())
 
-    with pytest.raises(trino.exceptions.TrinoAuthError) as exp:
-        oauth_test = OAuthTestReq("get", OAUTH_SERVER_URL_FORCE_FAIL)
-        token_req = oauth_test.get_token_request()
-        oauth(token_req)
+    redirect_server = f"https://coordinator/oauth2/initiate/{challenge_id}"
+    token_server = f"https://coordinator/oauth2/token/{challenge_id}"
```

Comment:
extract `redirect_server` and `token_server` to constants? those are repeated in a couple places.